### DISCUSSION TITLE
An implementation of MF2 with column-first selection

### DISCUSSION THREAD

**Discussion by eemeli** [unicode-org/message-format-wg#369] [Show and tell]
Following yesterday's discussion, I've updated my JS `Intl.MessageFormat` polyfill to provide two up-to-date implementations with the following selection method alternatives:

```sh
# first-match
npm install messageformat@4.0.0-3

# column-first
npm install messageformat@4.0.0-3.cf
```

The change in getting from the former to the latter is effectively commit [`f5d1bba`](https://github.com/messageformat/messageformat/commit/f5d1bba7b33b697eeb73bd9de1c01320f3d43bab), which changes the message value selection method from `matchSelectKey(key: string): boolean` to `selectKey(keys: Set<string>): string | null`. The actual selection algorithm is found in the `getPattern()` function in [`message-value/resolved-message.ts`](https://github.com/messageformat/messageformat/commit/f5d1bba7b33b697eeb73bd9de1c01320f3d43bab#diff-b29f80434e4ad10a57a8ee5161ee91beb0160d49a8334dcba20b7af440fd0082).

To play around with this locally, run one of the above shell commands and then create a new file `test.mjs` with these contents:
```js
import { MessageFormat } from "messageformat";

function test(msg, args) {
  const mf = new MessageFormat(msg, "en-US");
  const res = mf.resolveMessage(args);
  console.log(res.toString());
}

test(`
  let $x = {|1| :number}
  let $y = {|1| :number minimumFractionDigits=2}
  match {$x} {$y}
  when one one {{$x} is one, {$y} is one}
  when one 1 {{$x} is one, {$y} is exact}
  when 1 one {{$x} is exact, {$y} is one}
  when 1 1 {{$x} is exact, {$y} is exact}
  when * * {fallthrough}
`);
```
then run `node test.mjs` to observe the results. With `4.0.0-3`, the `when one 1` variant will be selected, while `4.0.0-3.cf` will select `when 1 1`. You may run the corresponding `npm install` command to switch from one to another.

To experiment with more complex rules while staying in English, the `:number` function supports the `type=ordinal` option to perform ordinal rather than cardinal selection.

CC @aphillips, @stasm, @mihnita

**Comment by macchiati** [comment]
Good idea to use ordinals in English (quite similar to Slavic cardinals)

**Comment by mihnita** [comment]
I've merged the Java implementation for best match selection.
It is in the `experimental` branch, in [`/experiments/mihnita/mf2_bestmatch`](https://github.com/unicode-org/message-format-wg/tree/experiments/experiments/mihnita/mf2_bestmatch)

---

It is a copy of the existing ICU4J implementation, with the classes moved to a different package to avoid conflicts.
Is still uses ICU to format dates, times, numbers, get plural rules, etc.

It is maven, so all you need to do is `mvn package`

---

I've added a dedicated test file ([`Mf2BestMatchTest`](https://github.com/unicode-org/message-format-wg/blob/experiments/experiments/mihnita/mf2_bestmatch/src/test/java/com/ibm/icu/message2bm/Mf2BestMatchTest.java)), 
 where the cases are sorted "the worst way possible" (`*` first, then `one` then `=1`).

And has an ordinal example:
```
match {$count :selectordinal}
 when * {The {$count}th place}
 when one {The {$count}st place}
 when two {The {$count}nd place}
 when few {The {$count}rd place}
 when 1 {Gold medal}
 when 2 {Silver medal}
 when 3 {Bronze medal}
```

---

The commits are cleanly separated to show the progress, from an empty maven project to the final version.
The only relevant commit is https://github.com/unicode-org/message-format-wg/commit/b06bba2ab4c58fd267855c63d9c5362223c3dfd0, that is the commit changing from first match to best match.

The changes needed are really minimal.

**Reply by eemeli** [reply]
Could you adjust the test suite to not depend on the system defaults? I'm getting quite a few errors that appear to be due to using `en-FI` as my system locale, and in a different timezone than expected.

**Reply by mihnita** [reply]
Thank you, good point.
Done.
M.

**Comment by eemeli** [comment]
Following #372, I've merged this into the JS polyfill in messageformat/messageformat#395.
